### PR TITLE
tty/usbacm: Block a reader, until all the data is read

### PR DESCRIPTION
**Pull request by MaciejPurski**
JIRA: DTR-324

## Description

This changes the behaviour of a blocking read. From now the reader should be blocked until all the data is read or an error happens. What's more the number of threads communicating with the USB host stack is reduced to one, in order to minimize the resources usage and ensure that urbs statuses are handled in an exact FIFO order.

## Motivation and Context

Those changes are needed in order to optimize read overhead by minimizing number of exchanged messages.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (imx6ull).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work (list the PRs, preferably in merge-order](https://github.com/phoenix-rtos/phoenix-rtos-usb/pull/26)).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: tty/usbacm/usbacm.c [hunk 1]
```diff
@@ -234,35 +208,126 @@ static void usbacm_put(usbacm_dev_t *dev)
...
 		return;
 
-	if (dev->flags != 0 && c->pipeid == dev->pipeBulkIN && c->err == 0) {
+	if (c->err != 0) {
+		/* Error, stop receiving */
+		mutexLock(dev->rxLock);
+		dev->rxrunning = 0;
+		if ((dev->flags & O_NONBLOCK) == 0) {
+			/* Stop the blocking read in progress */
+			condSignal(dev->rxCond);
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:224]: It's better to signal without mutex - condSignal does context switch

File: tty/usbacm/usbacm.c [hunk 2]
```diff
@@ -234,35 +208,126 @@ static void usbacm_put(usbacm_dev_t *dev)
...
+			memcpy(dev->rxptr, data, bytes);
+			dev->rxptr += bytes;
+			dev->rxbuflen -= bytes;
+
+			if (dev->rxbuflen > 0) {
+				retransfer = 1;
+			}
+			else {
+				dev->rxrunning = 0;
+				condSignal(dev->rxCond);
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:253]: It's better to signal without mutex - condSignal does context switch

File: tty/usbacm/usbacm.c [hunk 3]
```diff
@@ -234,35 +208,126 @@ static void usbacm_put(usbacm_dev_t *dev)
...
+			}
+		}
+		else {
+			dev->rxrunning = 0;
+			condSignal(dev->rxCond);
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:258]: It's better to signal without mutex - condSignal does context switch

File: tty/usbacm/usbacm.c [hunk 4]
```diff
@@ -234,35 +208,126 @@ static void usbacm_put(usbacm_dev_t *dev)
...
+		}
+		mutexUnlock(dev->rxLock);
+	}
+
+	if (retransfer) {
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:263]: MISRA

```suggestion
	if (retransfer != 0) {
```

File: tty/usbacm/usbacm.c [hunk 5]
```diff
@@ -234,35 +208,126 @@ static void usbacm_put(usbacm_dev_t *dev)
...
+	return 0;
+}
+
+
+static int usbacm_read(usbacm_dev_t *dev, char *data, size_t len)
+{
+	int ret;
+
+	if ((dev->flags & (O_RDONLY | O_RDWR)) == 0)
+		return -EPERM;
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:300]:
```suggestion
	if ((dev->flags & (O_RDONLY | O_RDWR)) == 0) {
		return -EPERM;
	}
```

File: tty/usbacm/usbacm.c [hunk 6]
```diff
@@ -302,62 +367,114 @@ static usbacm_dev_t *usbacm_devAlloc(void)
...
+		return -1;
+	}
 
-	if (dev->fifo == NULL) {
-		if ((fifo = malloc(sizeof(fifo_t) + RX_FIFO_SIZE)) == NULL)
-			return -ENOMEM;
+	for (i = 0; i < USBACM_N_URBS; i++) {
+		dev->urbs[i] = usb_urbAlloc(dev->pipeBulkIN, NULL, usb_dir_in, USBACM_BULK_SZ, usb_transfer_bulk);
+		if (dev->urbs[i] < 0) {
+			while (--i >= 0) {
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:382]: Two MISRA rules violated:
- modification of interator inside for loop
- condition with side-effect

File: tty/usbacm/usbacm.c [hunk 7]
```diff
@@ -302,62 +367,114 @@ static usbacm_dev_t *usbacm_devAlloc(void)
...
+static int _usbacm_openNonblock(usbacm_dev_t *dev)
+{
+	fifo_t *fifo;
+	int ret = 0;
 
-	dev->flags = flags;
-	for (i = 0; i < USBACM_N_URBS; i++)
-		dev->urbs[i] = usb_urbAlloc(dev->pipeBulkIN, NULL, usb_dir_in, USBACM_BULK_SZ, usb_transfer_bulk);
+	if ((fifo = malloc(sizeof(fifo_t) + RX_FIFO_SIZE)) == NULL)
+		return -ENOMEM;
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:401]:
```suggestion
	fifo = malloc(sizeof(fifo_t) + RX_FIFO_SIZE);
	if (fifo == NULL) {
		return -ENOMEM;
	}
```

File: tty/usbacm/usbacm.c [hunk 8]
```diff
@@ -302,62 +367,114 @@ static usbacm_dev_t *usbacm_devAlloc(void)
...
-	if (usbacm_init(dev) != EOK) {
-		fprintf(stderr, "usbacm: Init failed\n");
-		resourceDestroy(dev->fifoLock);
-		resourceDestroy(dev->fifoEmpty);
-		free(dev->fifo);
-		return -EINVAL;
+	if (usbacm_rxStart(dev) != 0) {
+		free(fifo);
+		dev->fifo = 0;
+		dev->rxrunning = 0;
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:412]: Hmm, condSignal not needed here?

File: tty/usbacm/usbacm.c [hunk 9]
```diff
@@ -302,62 +367,114 @@ static usbacm_dev_t *usbacm_devAlloc(void)
...
+			resourceDestroy(dev->rxLock);
+			return -EPERM;
+		}
+
+		if (flags & O_NONBLOCK) {
+			/* Start receving data */
+			if (_usbacm_openNonblock(dev) < 0) {
+				resourceDestroy(dev->rxCond);
+				resourceDestroy(dev->rxLock);
+				_usbacm_close(dev);
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:468]: Is resource destroy before _usbascm_close safe?

File: tty/usbacm/usbacm.c [hunk 10]
```diff
@@ -394,18 +511,24 @@ static void usbacm_msgthr(void *arg)
 			continue;
 		}
 
+		/* A device can be opened only by one process */
+		if (msg.type != mtOpen && (msg.pid != dev->clientpid)) {
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:515]:
```suggestion
		if ((msg.type != mtOpen) && (msg.pid != dev->clientpid)) {
```

Comment (agkaminski) [reply tty/usbacm/usbacm.c:515]: I guess it's ok for one process to open multiple times?

File: tty/usbacm/usbacm.c [hunk 11]
```diff
@@ -590,19 +713,13 @@ int main(int argc, char *argv[])
...
 	for (i = 0; i < USBACM_N_MSG_THREADS; i++) {
-		if ((ret = beginthread(usbacm_msgthr, 4, usbacm_common.stack[i], sizeof(usbacm_common.stack[i]), NULL)) != 0) {
-			fprintf(stderr, "usbacm: fail to beginthread ret: %d\n", ret);
-			return 1;
-		}
-	}
-
-	for (i = 0; i < USBACM_N_UMSG_THREADS; i++) {
-		if ((ret = beginthread(usbthr, 3, usbacm_common.ustack[i], sizeof(usbacm_common.ustack[i]), NULL)) != 0) {
+		if ((ret = beginthread(usbacm_msgthr, 3, usbacm_common.stack[i], sizeof(usbacm_common.stack[i]), NULL)) != 0) {
```
Comment (agkaminski) [inline tty/usbacm/usbacm.c:716]: right now beginthread does indeed return 0, but maybe we'll extend it's API to return TID on success instead
```suggestion
		ret = beginthread(usbacm_msgthr, 3, usbacm_common.stack[i], sizeof(usbacm_common.stack[i]), NULL);
		if (ret < 0) {
```

File: tty/usbacm/usbacm.c (hunk 8 above)
Comment (MaciejPurski) [reply tty/usbacm/usbacm.c:412]: No, this is a nonblocking case - no waiting on the conditional.

File: tty/usbacm/usbacm.c (hunk 9 above)
Comment (MaciejPurski) [reply tty/usbacm/usbacm.c:468]: `_usbacm_close` does not use neither `rxLock` nor `rxCond`.

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**